### PR TITLE
Refactor: adopt dependency injection pattern and extract cloud object.

### PR DESCRIPTION
**Pull request by MartinForReal**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
refactor: Code Refactoring 💎
Refactor: adopt dependency injection pattern and extract cloud object.
**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:
- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [x] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: pkg/azuredisk/azuredisk.go [hunk 1]
```diff
@@ -144,33 +135,21 @@ func newDriverV1(options *DriverOptions) *Driver {
...
-	driver.cloudConfigSecretName = options.CloudConfigSecretName
-	driver.cloudConfigSecretNamespace = options.CloudConfigSecretNamespace
-	driver.customUserAgent = options.CustomUserAgent
-	driver.userAgentSuffix = options.UserAgentSuffix
-	driver.useCSIProxyGAInterface = options.UseCSIProxyGAInterface
 	driver.enableDiskOnlineResize = options.EnableDiskOnlineResize
-	driver.allowEmptyCloudConfig = options.AllowEmptyCloudConfig
 	driver.enableAsyncAttach = options.EnableAsyncAttach
-	driver.enableListVolumes = options.EnableListVolumes
-	driver.enableListSnapshots = options.EnableListVolumes
```
Comment (MartinForReal) [inline pkg/azuredisk/azuredisk.go:156]: fixed the typo here.

File: pkg/azuredisk/azuredisk.go [hunk 2]
```diff
@@ -94,37 +97,25 @@ type hostUtil interface {
 // DriverCore contains fields common to both the V1 and V2 driver, and implements all interfaces of CSI drivers
 type DriverCore struct {
 	csicommon.CSIDriver
+	options                      DriverOptions
```
Comment (MartinForReal) [inline pkg/azuredisk/azuredisk.go:100]: Added a field to hold the options.

File: pkg/azuredisk/azuredisk.go [hunk 3]
```diff
@@ -144,33 +135,21 @@ func newDriverV1(options *DriverOptions) *Driver {
...
-	driver.enableWindowsHostProcess = options.EnableWindowsHostProcess
 	driver.getNodeIDFromIMDS = options.GetNodeIDFromIMDS
 	driver.enableOtelTracing = options.EnableOtelTracing
 	driver.shouldWaitForSnapshotReady = options.WaitForSnapshotReady
 	driver.volumeLocks = volumehelper.NewVolumeLocks()
 	driver.ioHandler = azureutils.NewOSIOHandler()
 	driver.hostUtil = hostutil.NewHostUtil()
-
+	driver.testingMock = options.TestingMock
+	driver.options = *options
```
Comment (MartinForReal) [inline pkg/azuredisk/azuredisk.go:152]: copy the options

**Review by MartinForReal — COMMENTED** [review]
added a field to hold the options.

File: pkg/azuredisk/azuredisk_test.go [hunk 4]
```diff
@@ -70,102 +69,6 @@ func TestCheckDiskCapacity(t *testing.T) {
 	assert.Equal(t, err, expectedErr)
 }
 
-func TestRun(t *testing.T) {
```
Comment (MartinForReal) [inline pkg/azuredisk/azuredisk_test.go:73]: test case is removed because I have no clue what scenario is covered here.

**Comment by MartinForReal** [comment]
<html>
<body>


/test pull-azuredisk-csi-driver-e2e-capz-windows-2022
--


<br class="Apple-interchange-newline">
</body>
</html>

File: pkg/azuredisk/azuredisk_test.go (hunk 4 above)
Comment (andyzhangx) [reply pkg/azuredisk/azuredisk_test.go:73]: this is for testing the cloud provider config `azure.json`

**Comment by MartinForReal** [comment]
<html>
<body>

/test pull-azuredisk-csi-driver-e2e-capz-multi-az
--


<br class="Apple-interchange-newline">
</body>
</html>

**Pull request closed without merging**